src/main/kotlin/com/cultureamp/eventsourcing/Upcast.kt

```diff
@@ -0,0 +1,24 @@
+package com.cultureamp.eventsourcing
+
+import kotlin.reflect.KClass
+import kotlin.reflect.full.functions
+
+interface Upcast<in M : EventMetadata, in E : DomainEvent, out OE : DomainEvent> {
+    fun upcast(event: E, metadata: M): OE
+}
+
+@Target(AnnotationTarget.CLASS)
+@MustBeDocumented
+annotation class UpcastEvent(val upcastType: KClass<out Upcast<*, *, *>>)
+
+fun UpcastEvent.upcasting(event: DomainEvent, metadata: EventMetadata): DomainEvent = this.upcastType
+    .functions
+    .filterIndexed { index, _ -> index == 0 }
```

Comment:
Why do we only look at the first function on the class?

Reply:
No needed anymore